### PR TITLE
{gcc6,gcc6Stdenv,gfortran6,gcj}: drop

### PR DESCRIPTION
**Pull request by emilazy**
## Description of changes

This release branch is 8½ years old and hasn’t received an update in 6 years. Nothing in the Nixpkgs tree uses it. We can simplify a lot of logic in the GCC and cc-wrapper derivations by removing this unsupported version.

I’m working on removing GCC 7 through 9 as well, as they cause complications for macOS by requiring an SDK that can target *very* old versions of macOS, but that’s blocked on CUDA stuff until probably after branch‐off. GCC 10 is also technically out of support, but seems like it will be required for CUDA for longer, and doesn’t cause any issues for Darwin.

Result of `nixpkgs-review` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>qmk</li>
    <li>qmk.dist</li>
  </ul>
</details>

I also built `hello` with `gcc7Stdenv` and `gcc8Stdenv` for `x86_64-linux` on `aarch64-linux` to check that cross still works. No rebuilds other than for old cross GCCs, and I can drop that commit if people don’t like it. I’d still appreciate people looking over the conditionals in case I messed things up for some non‐standard configuration.

cc @trofi @philiptaron @fabianhjr @RossComputerGuy

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by fabianhjr** [comment]
LGTM, will take a closer look during the weekend however 0 rebuilds is promising for not causing breakage.

**Comment by emilazy** [comment]
One rebuild, technically! :)

(Because we have exactly one use of a GCC < 9 cross-compiler in tree.)

File: pkgs/development/compilers/gcc/patches/default.nix [hunk 1]
```diff
@@ -266,35 +263,8 @@ in
   (./. + "/${majorVersion}/gcc8-asan-glibc-2.34.patch")
   (./. + "/${majorVersion}/0001-Fix-build-for-glibc-2.31.patch")
 ]
-++ optional ((is6 || is7) && targetPlatform.libc == "musl" && targetPlatform.isx86_32) (fetchpatch {
+++ optional (is7 && targetPlatform.libc == "musl" && targetPlatform.isx86_32) (fetchpatch {
   url = "https://git.alpinelinux.org/aports/plain/main/gcc/gcc-6.1-musl-libssp.patch?id=5e4b96e23871ee28ef593b439f8c07ca7c7eb5bb";
   sha256 = "1jf1ciz4gr49lwyh8knfhw6l5gvfkwzjy90m7qiwkcbsf4a3fqn2";
 })
-++ optional ((is6 || is7 || is8) && !atLeast9 && targetPlatform.libc == "musl") ./libgomp-dont-force-initial-exec.patch
```
Comment (emilazy) [inline pkgs/development/compilers/gcc/patches/default.nix:273]: Whoops, need to fix this tomorrow.

**Comment by emilazy** [comment]
Accidentally dropped a Musl patch and missed a stray 6 patch. Will address tomorrow.

File: pkgs/development/compilers/gcc/patches/default.nix [hunk 2]
```diff
@@ -266,35 +263,8 @@ in
...
-
-# Building libstdc++ with flat namespaces results in trying to link CoreFoundation, which
-# defaults to the impure, system location and causes the build to fail.
-++ optional (is6 && hostPlatform.isDarwin) ./6/libstdc++-disable-flat_namespace.patch
-
-## gcc 5.0 and older ##############################################################################
-
-++ optional (!atLeast6) ./parallel-bconfig.patch
-
+++ optional ((is7 || is8) && !atLeast9 && targetPlatform.libc == "musl") ./libgomp-dont-force-initial-exec.patch
```
Comment (emilazy) [inline pkgs/development/compilers/gcc/patches/default.nix:270]: Oh wait, the diff just moved it down here. Guess it's fine.

**Review by trofi — APPROVED** [review]

**Review by fabianhjr — APPROVED** [review]

**Comment by emilazy** [comment]
Fixed the dangling patch reference; this should be ready to land now.

**Comment by RossComputerGuy** [comment]
Result of `nixpkgs-review pr 341714` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>qmk</li>
    <li>qmk.dist</li>
  </ul>
</details>

**Review by RossComputerGuy — APPROVED** [review]

**Comment by emilazy** [comment]
Result of `nixpkgs-review pr 341714` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

I also verified the cross‐compilation test from the PR message on `x86_64-darwin` through Rosetta 2, as those compiler versions don’t support `aarch64-darwin` anyway.

**Comment by emilazy** [comment]
BTW, I’d appreciate `cc-wrapper` experts taking a look at https://github.com/NixOS/nixpkgs/pull/341715.

**Comment by Artturin** [comment]
Missed a gcc6 in `default-gcc-version` https://github.com/NixOS/nixpkgs/pull/343146

**Pull request merged**